config/configurator.go

```diff
@@ -0,0 +1,44 @@
+package config
+
+import "github.com/spf13/viper"
+
+type configurationLoader interface {
+	setConfig(name string, extension string, path string)
+	read() error
+	merge() error
+}
+
+type configurationReader interface {
+	getBool(key string) bool
+	getString(key string) string
+}
+
+type configurator interface {
+	configurationLoader
+	configurationReader
+}
+
+type ConfigProvider struct {
```

Comment:
From my understanding, we are injecting `ConfigProvider` into `Configuration`. `ConfigProvider` strongly depends on vendor library i.e. `viper`, and potentially can be replaced with something else.

If the assumption above is right, I have 2 questions:

1. Shall we rename `ConfigProvider` to `ViperProvider`, to specifically identify vendor dependency?
2. Is it the preferred way to handle dependency injection at the caller level, i.e. `config.Init(new(config.ViperProvider))`?

I have no problem with constructor injection, but if we follow the diagram and whenever we want to initialize a new component, we will have to follow the `package.Init(new(package.WhateverProvider))` pattern, I'm not sure if it's ideal. I know under the current design we can use:

- `package.Init(new(package.VendorProvider))` for "real" implementation,
- `package.Init(new(package.MockProvider))` for unit testing;

We can also create helper functions to use:

- `package.VendorSetup()` for "real" implementation,
- `package.MockSetup()` for unit testing;

And of course, we can also use DI container like [this](https://github.com/sarulabs/di). Personally I would prefer creating helper functions before introducing DI container, e.g. `config` initialization will look like:

- `config.ViperSetup()` or `config.SetupWithViper()` or something

I know you may not be a big fan of wrapper functions, I can compromise if you do insist. But I would at least writing the code in one line, as `provider` is only used in `Init` and an additional variable for each package initializer makes me feel a little big uneasy.

Reply:
Renamed `ConfigProvider` to `ViperProvider`.
at some point in future I'll add DI container and would pass container to the constructor instead of independent dependencies.

Reply:
Cool let's go with the current code for now, and whenever we are going to work on refactoring the 2nd component, let's do either the 2nd or 3rd way together.